src/data/sequence.rs

```diff
@@ -0,0 +1,727 @@
+        &self.data
+    }
+
+    /// Get owner at version.
+    pub fn owner_at(&self, version: impl Into<Version>) -> Option<&Owner> {
+        let version = to_absolute_version(version.into(), self.owners.len())?;
+        self.owners.get(version)
+    }
+
+    /// Returns history of all owners
+    pub fn owner_history(&self) -> Vec<Owner> {
+        self.owners.clone()
+    }
+
+    /// Get history of owners within the range of versions specified.
+    pub fn owner_history_range(&self, start: Version, end: Version) -> Option<Vec<Owner>> {
+        let range = to_absolute_range(start, end, self.owners.len())?;
+        Some(self.owners[range].iter().copied().collect())
+    }
+
+    /// Get access control at version.
+    pub fn access_list_at(&self, version: impl Into<Version>) -> Option<&C> {
+        let version = to_absolute_version(version.into(), self.access_list.len())?;
+        self.access_list.get(version)
+    }
+
+    /// Returns history of all access list states
+    pub fn access_list_history(&self) -> Vec<C> {
+        self.access_list.clone()
+    }
+
+    /// Get history of access list within the range of versions specified.
+    pub fn access_list_history_range(&self, start: Version, end: Version) -> Option<Vec<C>> {
+        let range = to_absolute_range(start, end, self.access_list.len())?;
+        Some(self.access_list[range].to_vec())
+    }
+
+    /// Set owner.
+    pub fn set_owner(&mut self, owner: Owner, expected_version: u64) -> Result<()> {
+        if owner.expected_data_version != self.expected_data_version() {
+            return Err(Error::InvalidSuccessor(self.expected_data_version()));
+        }
+        if owner.expected_access_list_version != self.expected_access_list_version() {
+            return Err(Error::InvalidPermissionsSuccessor(
+                self.expected_access_list_version(),
+            ));
+        }
+        if self.expected_owners_version() != expected_version {
+            return Err(Error::InvalidSuccessor(self.expected_owners_version()));
+        }
+        self.owners.push(owner);
+        Ok(())
+    }
+
+    /// Set access list.
+    /// The `AccessList` struct needs to contain the correct expected versions.
+    pub fn set_access_list(&mut self, access_list: &C, expected_version: u64) -> Result<()> {
+        if access_list.expected_data_version() != self.expected_data_version() {
+            return Err(Error::InvalidSuccessor(self.expected_data_version()));
+        }
+        if access_list.expected_owners_version() != self.expected_owners_version() {
+            return Err(Error::InvalidOwnersSuccessor(
+                self.expected_owners_version(),
+            ));
+        }
+        if self.expected_access_list_version() != expected_version {
+            return Err(Error::InvalidSuccessor(self.expected_access_list_version()));
+        }
+        self.access_list.push(access_list.clone()); // hmm... do we have to clone in situations like these?
```

Comment:
No the access_list argument shouldn't be a reference. That way the clone can be avoided.